libdevcore/Worker.h

```diff
@@ -29,74 +29,60 @@
-{
-	Fail,
-	Join,
-	Detach
-};
-
 enum class WorkerState
 {
-	Starting,
-	Started,
-	Stopping,
-	Stopped,
-	Killing
+	Running,			///< thread started and running
+	Stopped,			///< thread stopped, but still exists, ran be re-started
+	Terminated			///< thread does not exist
 };
 
-class Worker
+class Worker: boost::noncopyable
 {
 protected:
-	Worker(std::string const& _name = "anon", unsigned _idleWaitMs = 30): m_name(_name), m_idleWaitMs(_idleWaitMs) {}
-
-	/// Move-constructor.
-	Worker(Worker&& _m) { std::swap(m_name, _m.m_name); }
-
-	/// Move-assignment.
-	Worker& operator=(Worker&& _m) { std::swap(m_name, _m.m_name); return *this; }
+	/// Creates a new worker initially in the "Killed" state.
+	explicit Worker(std::string const& _name = "anon", unsigned _idleWaitMs = 30):
+		m_name(_name),
+		m_idleWaitMs(_idleWaitMs),
+		m_state{ WorkerState::Terminated },
+		m_requestedState{ WorkerState::Terminated }
+	{}
 
 	virtual ~Worker() { terminate(); }
 
-	/// Allows changing worker name if work is stopped.
-	void setName(std::string _n) { if (!isWorking()) m_name = _n; }
-
 	/// Starts worker thread; causes startedWorking() to be called.
 	void startWorking();
 	
 	/// Stop worker thread; causes call to stopWorking().
 	void stopWorking();
 
-	/// Returns if worker thread is present.
-	bool isWorking() const { Guard l(x_work); return m_state == WorkerState::Started; }
-	
+	/// Returns if worker thread is present and running.
+	bool isWorking() const { return m_state == WorkerState::Running; }
+	bool shouldStop() const { return m_requestedState != WorkerState::Running; }
+
 	/// Called after thread is started from startWorking().
 	virtual void startedWorking() {}
 	
 	/// Called continuously following sleep for m_idleWaitMs.
 	virtual void doWork() {}
 
-	/// Overrides doWork(); should call shouldStop() often and exit when true.
+	/// Calls/should call doWork repeatedly until shouldStop is true.
 	virtual void workLoop();
-	bool shouldStop() const { return m_state != WorkerState::Started; }
 	
 	/// Called when is to be stopped, just prior to thread being joined.
 	virtual void doneWorking() {}
 
-	/// Blocks caller into worker thread has finished.
-//	void join() const { Guard l(x_work); try { if (m_work) m_work->join(); } catch (...) {} }
-
 private:
 	/// Stop and never start again.
 	void terminate();
 
-	std::string m_name;
+	std::string const m_name;
 
-	unsigned m_idleWaitMs = 0;
+	unsigned const m_idleWaitMs = 0;
 	
-	mutable Mutex x_work;						///< Lock for the network existance.
-	std::unique_ptr<std::thread> m_work;		///< The network thread.
-	std::atomic<WorkerState> m_state = {WorkerState::Starting};
+	mutable Mutex x_work;						///< Lock for the worker thread.
+	std::unique_ptr<std::thread> m_work;		///< The worker thread.
+	std::atomic<WorkerState> m_state;			///< The current state of the thread, controlled by the thread.
+	std::atomic<WorkerState> m_requestedState;	///< The desired state of the thread, controlled from outside.
```

Comment:
Mutations of m_requestedState are binary, guarded, and reads are thread-safe. Atomic is unnecessary overkill. Worker doesn't provide, nor does it intend to provide real-time guarantees -- while(...) sleep() negates such a possibility (we do not know the value of any timer when state change is requested or enacted).